homeassistant/components/eq3btsmart/climate.py

```diff
@@ -20,79 +24,312 @@
 
+EQ_TO_STRING = {
+    eq3.Mode.Unknown: "unknown",
+    eq3.Mode.Closed: "closed",
+    eq3.Mode.Open: "open",
+    eq3.Mode.Auto: "auto",
+    eq3.Mode.Manual: "manual",
+    eq3.Mode.Away: "away",
+    eq3.Mode.Boost: "boost",
+}
+
+EQ_TO_HA_HVAC_ACTION = {
+    eq3.Mode.Open: CURRENT_HVAC_HEAT,
+    eq3.Mode.Closed: CURRENT_HVAC_IDLE,
+}
 
-DEVICE_SCHEMA = vol.Schema({vol.Required(CONF_MAC): cv.string})
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_MAC): cv.string,
+        vol.Optional(CONF_SENSOR): cv.entity_id,
+        vol.Optional(CONF_SENSOR_CONTROL, default=False): cv.boolean,
+        vol.Optional(CONF_MAX_TEMP): vol.Coerce(float),
+        vol.Optional(CONF_MIN_TEMP): vol.Coerce(float),
+        vol.Optional(CONF_PRECISION): vol.In(
+            [PRECISION_TENTHS, PRECISION_HALVES, PRECISION_WHOLE]
+        ),
+        vol.Optional(CONF_COLD_TOLERANCE, default=DEFAULT_TOLERANCE): vol.Coerce(float),
+        vol.Optional(CONF_HOT_TOLERANCE, default=DEFAULT_TOLERANCE): vol.Coerce(float),
+    }
+)
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
     {vol.Required(CONF_DEVICES): vol.Schema({cv.string: DEVICE_SCHEMA})}
 )
 
 SUPPORT_FLAGS = SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE
+SUPPORT_FLAGS_SENSOR_CONTROL = SUPPORT_TARGET_TEMPERATURE
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the eQ-3 BLE thermostats."""
     devices = []
 
     for name, device_cfg in config[CONF_DEVICES].items():
         mac = device_cfg[CONF_MAC]
-        devices.append(EQ3BTSmartThermostat(mac, name))
-
-    add_entities(devices, True)
-
-
-class EQ3BTSmartThermostat(ClimateEntity):
+        sensor_entity_id = device_cfg.get(CONF_SENSOR)
+        sensor_control = device_cfg.get(CONF_SENSOR_CONTROL)
+        min_temp = device_cfg.get(CONF_MIN_TEMP)
+        max_temp = device_cfg.get(CONF_MAX_TEMP)
+        precision = device_cfg.get(CONF_PRECISION)
+        cold_tolerance = device_cfg.get(CONF_COLD_TOLERANCE)
+        hot_tolerance = device_cfg.get(CONF_HOT_TOLERANCE)
+
+        devices.append(
+            EQ3BTSmartThermostat(
+                mac,
+                name,
+                sensor_entity_id,
+                sensor_control,
+                min_temp,
+                max_temp,
+                precision,
+                cold_tolerance,
+                hot_tolerance,
+            )
+        )
+
+    async_add_entities(devices, True)
+
+
+class EQ3BTSmartThermostat(ClimateEntity, RestoreEntity):
     """Representation of an eQ-3 Bluetooth Smart thermostat."""
 
-    def __init__(self, _mac, _name):
+    def __init__(
+        self,
+        _mac,
+        _name,
+        _sensor_entity_id,
+        _sensor_control,
+        _min_temp,
+        _max_temp,
+        _precision,
+        _cold_tolerance,
+        _hot_tolerance,
```

Comment:
The parameters should not be underscore prefixed as that is reserved for internals. A question, would it make more sense to pass the config and extract the data inside `__init__` instead of having a monster parameter listing? :-)

Reply:
I will rename the parameters.

I can't answer the question though, not that experienced here.
Let me know if I shall change the approach.